lua/telescope/finders/async_oneshot_finder.lua

```diff
@@ -1,100 +1,89 @@
-
-  local job
-
-  return setmetatable({
-    close = function()
-      if job then
-        job:close()
-      end
-    end,
-    results = results,
-  }, {
-    __call = function(_, prompt, process_result, process_complete)
-      if not job_started then
-        local job_opts = fn_command()
-
-        -- TODO: Handle writers.
-        -- local writer
-        -- if job_opts.writer and Job.is_job(job_opts.writer) then
-        --   writer = job_opts.writer
-        -- elseif job_opts.writer then
-        --   writer = Job:new(job_opts.writer)
-        -- end
-
-        stdout = LinesPipe()
-        job = async_job.spawn {
-          command = job_opts.command,
-          args = job_opts.args,
-          cwd = cwd,
-          env = env,
-
-          stdout = stdout,
-        }
-
-        job_started = true
-      end
-
-      if not job_completed then
-        if not vim.tbl_isempty(results) then
-          for _, v in ipairs(results) do
-            process_result(v)
-          end
-        end
-        for line in stdout:iter(false) do
-          num_results = num_results + 1
-
-          if num_results % await_count then
-            async.util.scheduler()
-          end
-
-          local v = entry_maker(line)
-          results[num_results] = v
-          process_result(v)
-        end
-
+local function finder_factory(opts)
+  local config = config_from_opts(opts)
+  -- TODO is it better to work on a copy?
+  local cached = config.results
+  local job, stdout
+
+  local function start_job_once()
+    start_job_once = function() end
+    local call = config.fn_command()
+    stdout = LinesPipe()
+    job = async_job.spawn {
+      command = call.command,
+      args = call.args,
+      cwd = config.cwd,
+      env = config.env,
+      stdout = stdout,
+    }
+  end
+
+  --[[
+  -- A finder is a callable with arguments (prompt, process_result, process_complete).
+  -- In general, the results it produces can depend on prompt.
+  -- For this one-shot finder here the results dont depend on prompt.
+  -- The user searches the results only based on fuzzy matching.
+  -- A finder callable should yield and not block too often.
+  -- Call process_result(entry) to add a new entry.
+  -- Exit early if process_result() returns true.
+  -- Always call process_complete() before exiting.
+  -- The picker relies on finder to reproduce all the same entries again when the prompt changes.
+  --]]
+  local function produce_entries(_, _, process_result, process_complete)
+    start_job_once()
+
+    for _, entry in ipairs(cached) do
+      -- TODO ok something is very wrong here, scheduler waits until we are on the main thread? then we always need it
+      -- or should process_result do it itself, because they want to call vim.api stuff
+      -- original did something like every 1000 iterations, but that makes no sense because you dont know how fast is your command
+      -- how does this relate to await_schedule()? ah the same. just an alias
```

Comment:
the reason the old one did every 1000 iterations was because it was expensive to do it every single time for something that might be called a million times.

We could do some testing to see how this fairs, but I'm hesitant to change it.

Reply:
When I tested it with the above script producing 1M entries, it was still super fast and very responsive, but my current machine is very up-to-date. I'm not saying we need it every single time, my problem is that I dont quite understand what's happening. Maybe you can help me:

Inside the `process_result` callback we do call `vim.api.nvim_*` functions. There are restrictions on when you can call nvim apis, right? Something like "It is an error to directly invoke vim.api functions in vim.loop callbacks.". I'm not sure in what context is the finder running in (an "libuv/plenary async" thread?) and if from that context I can call vim apis. Without the scheduler it sometimes failed trying to call the vim api.

What is `async.util.scheduler()` doing? Maybe then I get it. Plenary says "An async function that when awaited will await the scheduler to be able to call the api." Does that mean that my thread is yielding until it can be scheduled on the vim main thread and then I can call vim apis safely? That would mean only right after `scheduler()` becuase if something else yields on my thread I might already again be off the main thread?